Added some info about @Mod

Domochevsky opened the pull request:
Ideally more info can be added here, since I have not found a good source for most of these. I think some shenanigans can be done with "version" as well, defining wildcards and ranges of acceptable versions.

Domochevsky committed:
Added some info about @Mod

Ideally more info can be added here, since I have not found a good source for most of these.

Domochevsky committed:
Fixed some spelling and added more annotations

I think there are still more to explore (and to edit)

Domochevsky committed:
Some phrasing changes

Most for acceptedMinecraftVersions, to decouple it from Eclipse in particular.

diesieben07 commented on docs/gettingstarted/structuring.md:
```diff
@@ -25,6 +25,69 @@ What is `@Mod`?
 
 This is an annotation indicating to the Forge Mod Loader that the class is a Mod entry point. It contains various metadata about the mod. It also designates the class that will receive `@EventHandler` events. More information can be found at... (Coming Soon)
 
+All components of @Mod() are written inside the parantheses and chained together. There's only ever a single @Mod in your entire mod, sitting on top of your entry class.
```
You don't need to explain how annotations work, this is not a java tutorial. And there can be as many `@Mod` as you want.

Domochevsky replied:
Do I not, though? Where does the java tutorial end and the forge tutorial start?
Additionally: What would the purpose of multiple `@Mod` annotations be? Do they combine behind the scenes? 
This is not immediately apparent to whoever is new to modding. (I presume this description is aiming to help both the beginner and immediate modder.)

bonii-xx replied:
>Where does the java tutorial end

Before the forge tutorial starts. You don't see docs of libraries explaining java to you, do you? Java tutorial goes into java tutorial, Forge docs into forge docs.

diesieben07 replied:
About multiple `@Mod` annotations: Multiple `@Mod`s just mean that you have multiple mods.

Domochevsky replied:
> Before the forge tutorial starts.

You say that like there aren't lots of people who teach themselves java by modding for Minecraft. They need every leg up they can get. :)
(It is quite dangerous to assume that Common Knowledge is a thing.)

> Multiple @Mods just mean that you have multiple mods.

Are there any practical applications here to have more than one @Mod annotation?

diesieben07 replied:
Learning java by modding is a horrible idea.

About multiple mods: I can think of a few, but all I am saying is that you should not say "there is only ever one `@Mod`". I would put it like this: "There is one `@Mod` per mod.".

bonii-xx replied:
>>Before the forge tutorial starts.
>You say that like there aren't lots of people who teach themselves java by modding for >Minecraft. They need every leg up they can get. :)
>(It is quite dangerous to assume that Common Knowledge is a thing.)

That's what tutorials are for. This is a documentation.

Domochevsky replied:
> About multiple mods: I can think of a few, but all I am saying is that you should not say "there is only ever one @Mod". I would put it like this: "There is one @Mod per mod.".

That seems like a reasonable adjustment. It's useful in readability and hints at the idea that there might be more to it.

In general I'm trying to keep this information relevant to the task at hand, which includes minor dips into related topics. Whether or not that counts as doc or tut remains to be decided.

diesieben07 commented on docs/gettingstarted/structuring.md:
```diff
@@ -25,6 +25,69 @@ What is `@Mod`?
+Components:
+
+`modid` [Required]
+This identifies your mod uniquely to Forge and should probably be the first thing you enter. It's only visible internally.
```
No, it is not. The modId is public.

Domochevsky replied:
Hm, phrasing maybe. "Only used internally and of no relevance to the end user". Of course they can see it when they look into the files.

bonii-xx replied:
It's not only used internally, other mods may need to know it too for dependencies or IMCs (or whatever they do)

Domochevsky replied:
Hm, good point. Adding that as well.

diesieben07 commented on docs/gettingstarted/structuring.md:
```diff
@@ -25,6 +25,69 @@ What is `@Mod`?
+`modid` [Required]
+This identifies your mod uniquely to Forge and should probably be the first thing you enter. It's only visible internally.
+
+Example: @Mod(modid="examplemod")
```
Again, no need to explain how annotations work.

Domochevsky replied:
That is presuming that people know what annotations are beforehand. Which I find dangerous. :)

diesieben07 replied:
If people don't know what annotations are they can go to the [java tutorial about them](https://docs.oracle.com/javase/tutorial/java/annotations/index.html).

Domochevsky replied:
Sensible. Adding a link to it with a "wanna know more?".

diesieben07 replied:
I mean, you could. But I don't see why, the title page for this documentation explicitly states that it is **not** a java tutorial.

diesieben07 commented on docs/gettingstarted/structuring.md:
```diff
@@ -25,6 +25,69 @@ What is `@Mod`?
+
+`version` [Required]
+This is the version of your mod and gets both publically displayed and internally used to detect version mismatches.
+You can write pretty much whatever in here, but it's generally a good idea to have some sort of structure. Even if it's something as simple as a single number that increases every time you update your mod.
```
There is already an article about versioning, you should link to that instead.

Domochevsky replied:
Additionally, you mean. But yeah, sensible.

diesieben07 commented on docs/gettingstarted/structuring.md:
```diff
@@ -25,6 +25,69 @@ What is `@Mod`?
+Example: [TODO]
+
+`canBeDeactivated` [Optional]
+This lets Forge know whether or not your mod can be safely turned off/on during gameplay. There's a menu option for enabling/disabling mods in the main menu, which uses this. Unlike most other annotation options this one takes a boolean instead of a string, so it's without quotes.
```
Again, no need to explain basic java concepts.

Akxe replied:
It is to be removed in 1.13

diesieben07 commented on docs/gettingstarted/structuring.md:
```diff
@@ -25,6 +25,69 @@ What is `@Mod`?
+Example: @Mod(canBeDeactivated=true)
+
+`clientSideOnly` [Optional]
+This prevents the mod from being loaded on the server at all, no matter what. If this is set together with the option below the universe will explode. Or the game crashes. One or the other.
```
This is a documentation, be serious. Test what actually happens (hint: it will crash).

Domochevsky replied:
Mh, sensible enough. Gonna remove the humor from the next edit.

diesieben07 commented on docs/gettingstarted/structuring.md:
```diff
@@ -25,6 +25,69 @@ What is `@Mod`?
+Example: @Mod(clientSideOnly=true)
+
+`serverSideOnly` [Optional]
+The opposite of the option above. This prevents the mod from being loaded on client-side. Beware: This also includes the internal server in singleplayer.
```
Maybe link to the article about sides here to explain the difference between logical and physical sides.

Domochevsky replied:
Article about sides... ah, ok yeah, makes sense.

diesieben07 commented on docs/gettingstarted/structuring.md:
```diff
@@ -25,6 +25,69 @@ What is `@Mod`?
+Example: @Mod(serverSideOnly=true)
+
+`dependencies` [Optional]
+This defines what other mods yours depends on. Possibly also the reverse. Multiple entries can be made here, separated by a ; (What is this called?).
```
Semicolon.

Domochevsky replied:
Thanks.

diesieben07 commented on docs/gettingstarted/structuring.md:
```diff
@@ -25,6 +25,69 @@ What is `@Mod`?
+
+`dependencies` [Optional]
+This defines what other mods yours depends on. Possibly also the reverse. Multiple entries can be made here, separated by a ; (What is this called?).
+[TODO: There seems to be some additional constructs and structures here that someone needs to expand on. Is there something other than "required-after"?]
```
Yes, `after`, `before`, `required-after`, `required-before`. The ones without "required" are soft-dependencies that simply control ordering.
You should probably explain all this here (including the version ranges) instead of slapping down an example.

Domochevsky replied:
I was planning to, yeah, once I had the necessary information together.

Do these words all behave like I presume they do?

**before** - loaded before the specified mod
_after_ - loaded after the specified mod
_required-after_ - loaded after the specified mod, which HAS to be there
_required-before_ - loaded before the specified mod, which HAS to be there

diesieben07 replied:
Yes.

Domochevsky replied:
Gotcha. Can you explain version ranges in further detail so I can add that as well?

diesieben07 commented on docs/gettingstarted/structuring.md:
```diff
@@ -25,6 +25,69 @@ What is `@Mod`?
+Example: @Mod(dependencies="required-after:Forge@[12.16.0.1859,);")
+
+`acceptedMinecraftVersions` [Optional]
+This tells Forge what versions of Minecraft your mod runs on. By default it's set to "whatever I'm currently building my mod against in my editor of choice". Tends to be quite specific. Sub-versions are not naturally included, so keep an eye out there.
```
The default is actually set to "don't care, all versions". But ForgeGradle sets this property when building your mod.

Domochevsky replied:
Mh, interesting. Gonna add that tidbit to the next edit as well.

Domochevsky commented on docs/gettingstarted/structuring.md:
```diff
@@ -25,6 +25,69 @@ What is `@Mod`?
+Example: [TODO]
+
+`canBeDeactivated` [Optional]
+This lets Forge know whether or not your mod can be safely turned off/on during gameplay. There's a menu option for enabling/disabling mods in the main menu, which uses this. Unlike most other annotation options this one takes a boolean instead of a string, so it's without quotes.
```
Which concepts would those be?

Also: Mod disabling... it doesn't work? Like, at all? o_O

diesieben07 replied:
> Which concepts would those be?

What a boolean is and how you put them into an annotation.
> Also: Mod disabling... it doesn't work? Like, at all? o_O

Last I checked, yes.

Domochevsky replied:
Curious. Is there an plan for when that's gonna be functional?

diesieben07 replied:
Actually there seems to be some support for it through the `runtimeDisableMod` method, but it's never used.

tterrag1098 replied:
Yeah I'd remove this bit completely. Or just mention that it's not used.

Domochevsky committed:
More phrasing changes and info added

Added a couple data points where mentioned in comments from the previous edit.
There's a couple more things to mess with, as marked by the TODOs. (Like, how do version ranges work? How are they written and why?)

diesieben07 commented on docs/gettingstarted/structuring.md:
```diff
@@ -30,7 +30,8 @@ All components of @Mod() are written inside the parantheses and chained together
 
 `modid` [Required]
-This identifies your mod uniquely to Forge and should probably be the first thing you enter. It's only visible internally.
+This identifies your mod uniquely to Forge and should probably be the first thing you enter. It's only used internally, but can be seen by users who dig around. Other mods also require this ID if they want to find each other. (APIs for example.)
```
Nitpicky, but still: periods go after the parentheses, not before: 
"each other (APIs for example)."

diesieben07 commented on docs/gettingstarted/structuring.md:
```diff
@@ -30,7 +30,8 @@ All components of @Mod() are written inside the parantheses and chained together
 `modid` [Required]
-This identifies your mod uniquely to Forge and should probably be the first thing you enter. It's only visible internally.
+This identifies your mod uniquely to Forge and should probably be the first thing you enter. It's only used internally, but can be seen by users who dig around. Other mods also require this ID if they want to find each other. (APIs for example.)
+If you want to know more about annotations in general you can check out [Oracle](https://docs.oracle.com/javase/tutorial/java/annotations/index.html)'s docs on it.
```
You are linking to the docs, not to Oracle. Hence the link should be on either "docs" or "Oracle's docs" as a whole.

Domochevsky replied:
Oh yeah. Adjusting that.

diesieben07 commented on docs/gettingstarted/structuring.md:
```diff
@@ -42,6 +43,7 @@ Example: @Mod(modid="examplemod", name="Mad Mod For Kool Kids")
 `version` [Required]
 This is the version of your mod and gets both publically displayed and internally used to detect version mismatches.
 You can write pretty much whatever in here, but it's generally a good idea to have some sort of structure. Even if it's something as simple as a single number that increases every time you update your mod.
+Some more info about that can be found on the [left-hand menu](https://mcforge.readthedocs.io/en/latest/conventions/versioning/).
```
Why not "found in the article about versioning"?

diesieben07 commented on docs/gettingstarted/structuring.md:
```diff
@@ -58,23 +60,37 @@ Beware: This should NEVER be true for mods that add things to the game, like ite
-The opposite of the option above. This prevents the mod from being loaded on client-side. Beware: This also includes the internal server in singleplayer.
+The opposite of the option above. This prevents the mod from being loaded on client-side. 
+Beware: This also includes the internal server in singleplayer.
+What sides are and what their deal is can be read up on via the [left-hand menu](https://mcforge.readthedocs.io/en/latest/concepts/sides/).
```
See above. "left-hand menu" is just as bad of a link text as "click here".

diesieben07 commented on docs/gettingstarted/structuring.md:
```diff
@@ -58,23 +60,37 @@ Beware: This should NEVER be true for mods that add things to the game, like ite
+`before` - Loaded before the specified mod-
+`after` - Loaded after the specified mod-
+`required-after` - Loaded after the specified mod, which HAS to be there.
+`required-before` - Loaded before the specified mod, which HAS to be there.
```
We use **bold** to emphasize, not UPPERCASE.

Domochevsky committed:
More phrasing adjustments

tterrag1098 commented:
I believe there is enough info here to deserve its own page

tterrag1098 commented on docs/gettingstarted/structuring.md:
```diff
@@ -25,6 +25,84 @@ What is `@Mod`?
+This identifies your mod uniquely to Forge and should probably be the first thing you enter. It's only used internally, but can be seen by users who dig around. Other mods also require this ID if they want to find each other (APIs for example).
+If you want to know more about annotations in general you can check out [Oracle's docs](https://docs.oracle.com/javase/tutorial/java/annotations/index.html) on it.
+
+Example: `@Mod(modid="examplemod")`
```
These examples are pointless. Again this is not a java tutorial. Simply explaining the different annotation fields is sufficient.

Domochevsky replied:
Is there any damage or overall decrease of quality of the doc to include these examples?

kimfy replied:
Personally, as someone who is using wikis a lot; having to read through basics of the programming language I'm developing in is becoming a nuisance. I'm here to learn about your API, not the language I'm using. I know complete beginners probably appreciate it, but for all the people that are already familiar with the language it's not adding anything but more to read. Less is more in this case, in my opinion.

LexManos replied:
We are not here to teach java 101 Do not do it.

tterrag1098 commented on docs/gettingstarted/structuring.md:
```diff
@@ -25,6 +25,84 @@ What is `@Mod`?
+
+`version` [Required]
+This is the version of your mod and gets both publically displayed and internally used to detect version mismatches.
+You can write pretty much whatever in here, but it's generally a good idea to have some sort of structure. Even if it's something as simple as a single number that increases every time you update your mod.
```
"pretty much whatever" is not a very professional phrase

Domochevsky replied:
I did not aim to be professional.

tterrag1098 commented on docs/gettingstarted/structuring.md:
```diff
@@ -25,6 +25,84 @@ What is `@Mod`?
+Example: `@Mod(modid="examplemod", name="Mad Mod For Kool Kids" version="1.radical")`
+
+`acceptableRemoteVersions` [Optional]
+This lets Forge know what Client-side versions your mod is ok with. The wildcard * (a star) means that your mod is fine with ANY version on the other end, even nothing at all. You can use this for Server-side only mods.
```
"is ok with" -> "accepts when run on a server"

Domochevsky replied:
Reasonable.

tterrag1098 commented on docs/gettingstarted/structuring.md:
```diff
@@ -25,6 +25,84 @@ What is `@Mod`?
+
+`clientSideOnly` [Optional]
+This prevents the mod from being loaded on the server at all, no matter what. 
+Beware: If this is set together with the option below the game will crash, no questions asked.
```
This information could be put in an admonition after these two sections. For instance:

!!! note
clientSideOnly and serverSideOnly must not both be true, for obvious reasons.

tterrag1098 commented:
Fix the remaining comments, get rid of all the "java tutorial" stuff, and generally make the wording a bit more professional. Anything more nit-picky I'll do myself.

Domochevsky committed:
Wording adjustments

I continue to be ambivalent on whether or not "java tutorial" stuff should be left out. It doesn't seem to be harmful to this doc to mention these things. If anything they increase general understanding.

tterrag1098 commented:
>I continue to be ambivalent on whether or not "java tutorial" stuff should be left out. It doesn't seem to be harmful to this doc to mention these things. If anything they increase general understanding.

It's not your decision. It is well documented (on the README no less) that this is *not* a java tutorial.

Domochevsky commented:
Hm, fair enough then. Which parts would you like to be removed?

tterrag1098 commented:
They've already been discussed above.

williewillus commented on docs/gettingstarted/structuring.md:
```diff
@@ -25,6 +25,86 @@ What is `@Mod`?
+Of special note: If clientSideOnly and serverSideOnly are both set the game will crash, no questions asked.
+
+`dependencies` [Optional]
+This defines what other mods yours depends on. Possibly also the reverse. Multiple entries can be made here, separated by a semicolon.
```
"Possibly also the reverse"

No fluffy indecisive language - these are docs for beginners. Find out if it does or not :P

williewillus commented on docs/gettingstarted/structuring.md:
```diff
@@ -25,6 +25,86 @@ What is `@Mod`?
+
+`updateJSON` [Optional]
+This does... something. JSON-related, likely.
+[TODO: Someone explain this. Certainly ain't gonna be me.]
```
Points to the URL of a JSON to use in Forge's lightweight update-checker system

williewillus commented on docs/gettingstarted/structuring.md:
```diff
@@ -25,6 +25,86 @@ What is `@Mod`?
+
+Using these keywords you can establish some basic ordering for your mod, in the grand scheme of things.
+
+[TODO: Version ranges seem to be a thing as well. How do?]
```
MODID@[min,max]

Follows standard mathematical notation - [ and ] describe inclusive range, () describe exclusive range.
min or max (but not both) can be omitted for an unbounded version in that direction. For example "[1.9,)" matches all 1.9+ (but that's bad please don't do that)

williewillus commented on docs/gettingstarted/structuring.md:
```diff
@@ -25,6 +25,86 @@ What is `@Mod`?
+This tells Forge what versions of Minecraft your mod runs on. By default it's set to "don't care", but ForgeGradle defaults to "whatever I'm currently building my mod against" when you compile your mod. 
+Beware: ForgeGradle is quite specific. Sub-versions are not naturally included, so keep an eye out there.
+
+[TODO: Multiple versions and constructs can be added here. Explain them. "[1.9,1.10)" seems to be a valid construct for 1.9.*, including the bracket/parantheses. What makes it work?]
```
see above

williewillus commented on docs/gettingstarted/structuring.md:
```diff
@@ -25,6 +25,86 @@ What is `@Mod`?
+
+`acceptableRemoteVersions` [Optional]
+This lets Forge know what Client-side versions your mod accepts when run on a server. The wildcard * (a star) means that your mod is fine with ANY version on the other end, even nothing at all. You can use this for Server-side only mods.
+[TODO: Someone add here how the versions can be structured. I THINK you can set minimum/maximum versions?]
```
Same range system as I commented below

williewillus commented on docs/gettingstarted/structuring.md:
```diff
@@ -25,6 +25,86 @@ What is `@Mod`?
 
 This is an annotation indicating to the Forge Mod Loader that the class is a Mod entry point. It contains various metadata about the mod. It also designates the class that will receive `@EventHandler` events. More information can be found at... (Coming Soon)
 
+All components of @Mod() are written inside the parantheses and chained together. There's typically only one @Mod in your mod as a whole, sitting on top of your entry class.
```
clarify "chained together" or take it out.

Like said below we assume the modder knows how to specify annotation parameters

williewillus commented on docs/gettingstarted/structuring.md:
```diff
@@ -25,6 +25,86 @@ What is `@Mod`?
+Components:
+
+`modid` [Required]
+This identifies your mod uniquely to Forge and should probably be the first thing you enter. It's only used internally, but can be seen by users who dig around. Other mods also require this ID if they want to find each other (APIs for example).
```
"and should probably be the first thing you enter. It's only used internally, but can be seen by users who dig around"

Too much fluff.

Something like
"Uniquely identifies your mod to FML and all other mods. Should not be changed after you release your mod, as FML would then consider it a new, separate mod"
